kernel/src/engine/default/file_handler.rs

```diff
@@ -86,19 +90,64 @@ pub struct FileStream<F: FileOpener> {
     /// (before reaching the limit) and returns a batch iterator. If the file reader
     /// is not capable of limiting the number of records in the last batch, the file
     /// stream will take care of truncating it.
-    file_reader: F,
+    file_reader: Box<dyn FileOpener>,
     /// The stream state
     state: FileStreamState,
     /// Describes the behavior of the `FileStream` if file opening or scanning fails
     on_error: OnError,
 }
 
-impl<F: FileOpener> FileStream<F> {
+impl FileStream {
+    pub fn new_file_data_read_iterator<E: TaskExecutor>(
```

Comment:
Is there a reason you have this in the `impl` for `FileStream`? 

It does make sense to hoist it _somewhere_ and not duplicate the code in json/parquet.rs, but probably just putting it in a stand-alone function in here would make more sense, since it's not really part of FileStream (it just uses one).

Reply:
Main reason was, otherwise there's no obvious connection between `FileStream` and all the supporting infra, vs. this seemingly random stand-alone function? I debated calling it `FileStream::new_async_iterator`, would that make more sense?

Reply:
I do think `new_async_iterator` makes more sense. I generally feel like methods in an `impl` should either operator on, or produce the type the are implemented for, which we don't really here. But in some sense what we return is a "wrapped" version of `FileStream` so that's probably okay.